### DISCUSSION TITLE
nrf52840 Dongle -- Not enough memory?

### DISCUSSION THREAD

**Discussion by rahav** [contiki-ng/contiki-ng#2726] [Q&A] [answered]
I would like to inquire about the current state of Contiki compatibility with Nordic nRF52840 platform. There was a compelling paper [advocating for Contiki](https://www.sciencedirect.com/science/article/pii/S2352711022000620), and Contiki has indeed proven to be a valuable tool in our research endeavors, especially Contiki 4.9. I have recently been recommending Nordic products in multiple contexts and I am on the verge of making a significant procurement decision from Nordic for our research lab. However, it appears that there are frequent challenges arising with Nordic. I would appreciate any insights or updates on the situation. Please see below for the most recent issue:

Code that used to perfectly run on the Dongle, stopped running. It runs with no issues on the DK, so we started to suspect memory requirements. However, checking the executable suggest that it should fit on the Dongle:
````
arm-none-eabi-size build/nrf52840/dongle/device.nrf52840 
   text	   data	    bss	    dec	    hex	filename
 100740	   2224	  17092	 120056	  1d4f8	build/nrf52840/dongle/device.nrf52840
````

After lots of digging we found out that the code stopped working on the dongle when Contiki Shell module was added to the project. Removing Shell yields:
````
arm-none-eabi-size build/nrf52840/dongle/device.nrf52840 
   text	   data	    bss	    dec	    hex	filename
  91564	   2044	  16948	 110556	  1afdc	build/nrf52840/dongle/device.nrf52840
````

Surprisingly, after this change the code runs on the Dongle. For reference, we determine that the code is running by viewing its output: `sudo picocom -fh -b 115200 --imap lfcrlf /dev/ttyACM0`

To eliminate any concerns that Contiki Shell is the problem, we kept it removed, and changed RPL routing from Lite to Classic. In this configuration the code does not run on the dongle as well. Thus, we now suspect memory requirements with higher confidence, but the memory of the code reported above are below what is available on the dongle.

Furthermore, there is no warning during `make TARGET=nrf52840 BOARD=dongle PORT=/dev/ttyACM0 device.dfu-upload` the upload, so the code does appear to fit.

Can someone please help us to determine how to resolve this issue?

**Comment by pjonsson** [comment]
From what I can see on the Nordic homepage, it's the same amount of memory on the dongle and DK.

Since removing something improved the state of things, my guess is that there is memory corruption somewhere in the code that is run, and not having the shell present made the corruption overwrite something less essential.

If the code can run on the native platform, I would suggest starting in that end and see if Valgrind says anything interesting.

**Reply by rahav** [reply]
Thank you @pjonsson, this is a wise advice. I will try as you suggested.

For the system to run, I need a few nodes to exchange messages via 802.15.4. What is the proper way to enable communication native nodes? Should it be done inside Contiki containers or at the Ubuntu host OS?

I attempted to run two hello-world nodes outside of a container, using `sudo ./hello-world.native` but it was able to open a `tun` interface only for the first node:

````
sudo ./hello-world.native 
opened tun device ``/dev/tun0''
tun0: flags=4305<UP,POINTOPOINT,RUNNING,NOARP,MULTICAST>  mtu 1500
        inet 127.0.1.1  netmask 255.255.255.255  destination 127.0.1.1
        inet6 fe80::3d4e:25f6:2543:411b  prefixlen 64  scopeid 0x20<link>
        inet6 fd00::1  prefixlen 64  scopeid 0x0<global>
        unspec 00-00-00-00-00-00-00-00-00-00-00-00-00-00-00-00  txqueuelen 500  (UNSPEC)
        RX packets 0  bytes 0 (0.0 B)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 0  bytes 0 (0.0 B)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0

[INFO: Main      ] Starting Contiki-NG-288b134d6-dirty
[INFO: Main      ] - Routing: RPL Lite
[INFO: Main      ] - Net: tun6
[INFO: Main      ] - MAC: nullmac
[INFO: Main      ] - 802.15.4 PANID: 0xabcd
[INFO: Main      ] - 802.15.4 Default channel: 26
[INFO: Main      ] Node ID: 1800
[INFO: Main      ] Link-layer address: 0102.0304.0506.0708
[INFO: Main      ] Tentative link-local IPv6 address: fe80::302:304:506:708
[INFO: Native    ] Added global IPv6 address fd00::302:304:506:708
Hello, world
````

but the second:
````
sudo ./hello-world.native 
[WARN: Tun6      ] Failed to open tun device (you may be lacking permission). Running without network.
[INFO: Main      ] Starting Contiki-NG-288b134d6-dirty
[INFO: Main      ] - Routing: RPL Lite
[INFO: Main      ] - Net: tun6
[INFO: Main      ] - MAC: nullmac
[INFO: Main      ] - 802.15.4 PANID: 0xabcd
[INFO: Main      ] - 802.15.4 Default channel: 26
[INFO: Main      ] Node ID: 1800
[INFO: Main      ] Link-layer address: 0102.0304.0506.0708
[INFO: Main      ] Tentative link-local IPv6 address: fe80::302:304:506:708
[INFO: Native    ] Added global IPv6 address fd00::302:304:506:708
Hello, world
````

**Reply by rahav** [reply, marked as answer]
The ARM compiler that comes with Contiki `nrf52840` platform (`TARGET=nrf52840`) likely has a bug that causes memory corruption, when your code has certain ROM sizes or organizations. If a mote does not output any LOG lines, you should suspect this bug.

To fix this you need to change your ROM size a little. Since code instructions go into ROM, one way to do this is adding a benign statement. Sometimes even a single `printf("\n");` or `LOG_WARN("\n");` suffice.

The newer, `nrf` platform (`TARGET=nrf`) do not seem to have this bug.

Thank you @pjonsson for your guidance throughout this investigation.
